src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonPropertyInfo.cs

```diff
@@ -1019,41 +1020,83 @@ public static void JsonIgnoreConditionIsCorrectlyTranslatedToShouldSerializeDele
 
                         Assert.Null(property.Get);
                         Assert.Null(property.Set);
                         break;
                     case JsonIgnoreCondition.WhenWritingDefault:
-                        Assert.NotNull(property.ShouldSerialize);
-                        Assert.False(property.ShouldSerialize(null, 0));
-                        Assert.True(property.ShouldSerialize(null, 1));
-                        Assert.True(property.ShouldSerialize(null, -1));
+                        TestIntIgnoreWhenWritingDefault();
                         break;
                     case JsonIgnoreCondition.WhenWritingNull:
-                        Assert.NotNull(property.ShouldSerialize);
-                        Assert.False(property.ShouldSerialize(null, null));
-                        Assert.True(property.ShouldSerialize(null, ""));
-                        Assert.True(property.ShouldSerialize(null, "asd"));
+                        TestStringIgnoreWhenWritingDefault();
                         break;
                     case JsonIgnoreCondition.Never:
-                        Assert.NotNull(property.ShouldSerialize);
-                        Assert.True(property.ShouldSerialize(null, null));
-                        Assert.True(property.ShouldSerialize(null, ""));
-                        Assert.True(property.ShouldSerialize(null, "asd"));
+                        Assert.Null(property.ShouldSerialize);
                         break;
+
+                    case null:
+                        switch (defaultIgnoreCondition)
+                        {
+                            case JsonIgnoreCondition.WhenWritingDefault:
+                                if (property.PropertyType == typeof(int))
+                                {
+                                    TestIntIgnoreWhenWritingDefault();
+                                }
+                                else
+                                {
+                                    Assert.Equal(typeof(string), property.PropertyType);
+                                    TestStringIgnoreWhenWritingDefault();
+                                }
+                                break;
+
+                            case JsonIgnoreCondition.WhenWritingNull:
+                                if (property.PropertyType == typeof(int))
+                                {
+                                    Assert.Null(property.ShouldSerialize);
+                                }
+                                else
+                                {
+                                    Assert.Equal(typeof(string), property.PropertyType);
+                                    TestStringIgnoreWhenWritingDefault();
+                                }
+                                break;
+
+                            default:
+                                Assert.Null(property.ShouldSerialize);
```

Comment:
I'd expect this to be true for all cases above but serialization behave exactly as defined by current ShouldSerialize

Reply:
Merge with https://github.com/dotnet/runtime/pull/71908#discussion_r917687312